libs/vkd3d/device.c

```diff
@@ -5106,17 +5121,109 @@ static HRESULT STDMETHODCALLTYPE d3d12_device_OpenSharedHandleByName(d3d12_devic
+            {
+                struct d3d12_heap *heap_object = impl_from_ID3D12Heap(heap_iface);
+
+                memory = heap_object->allocation.device_allocation.vk_memory;
+                priority = heap_object->priority;
+                heap_object->evicted = false;
+
+                ID3D12Heap_Release(heap_iface);
+            }
+            else if (SUCCEEDED(ID3D12DeviceChild_QueryInterface(object, &IID_ID3D12Resource, (void**)&resource_iface)))
+            {
+                struct d3d12_resource *resource_object = impl_from_ID3D12Resource(resource_iface);
+
+                memory = (resource_object->flags & VKD3D_RESOURCE_COMMITTED) ?
+                    resource_object->mem.device_allocation.vk_memory : VK_NULL_HANDLE;
+                priority = resource_object->priority;
+                resource_object->evicted = false;
+
+                ID3D12Resource_Release(resource_iface);
+            }
+
+            if (memory)
+            {
+                VK_CALL(vkSetDeviceMemoryPriorityEXT(device->vk_device, memory, vkd3d_convert_to_vk_prio(priority)));
+            }
+        }
+    }
 
     return S_OK;
 }
 
+static HRESULT STDMETHODCALLTYPE d3d12_device_EnqueueMakeResident(d3d12_device_iface *iface,
+        D3D12_RESIDENCY_FLAGS flags, UINT num_objects, ID3D12Pageable *const *objects,
+        ID3D12Fence *fence_to_signal, UINT64 fence_value_to_signal)
+{
+    /* note: we ignore flags/D3D12_RESIDENCY_FLAG_DENY_OVERBUDGET; it involves
+       knowing if the app will be made over-budget.  We act as if it won't.  Could perhaps
+       use VK_EXT_memory_budget but don't have an app in-hand that clearly cares. */
+    d3d12_device_MakeResident(iface, num_objects, objects);
+
+    /* we don't block anyway - signal the fence immediately */
+    return ID3D12Fence_Signal(fence_to_signal, fence_value_to_signal);
+}
+
 static HRESULT STDMETHODCALLTYPE d3d12_device_Evict(d3d12_device_iface *iface,
         UINT object_count, ID3D12Pageable * const *objects)
 {
-    FIXME_ONCE("iface %p, object_count %u, objects %p stub!\n",
-            iface, object_count, objects);
+    struct d3d12_device *device = impl_from_ID3D12Device(iface);
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+
+    if (device->vk_info.EXT_pageable_device_local_memory)
+    {
+        uint32_t i;
+        
+        for (i = 0; i < object_count; i++)
+        {
+            ID3D12Heap *heap_iface;
+            ID3D12Resource *resource_iface;
+            ID3D12DeviceChild *object = (ID3D12DeviceChild*)objects[i];
+            VkDeviceMemory memory = VK_NULL_HANDLE;
+
+            if (SUCCEEDED(ID3D12DeviceChild_QueryInterface(object, &IID_ID3D12Heap, (void**)&heap_iface)))
+            {
+                struct d3d12_heap *heap_object = impl_from_ID3D12Heap(heap_iface);
+
+                memory = heap_object->allocation.device_allocation.vk_memory;
```

Comment:
Same concerns here as for MakeResident.

Reply:
It seems useful to have a VKD3D_RESOURCE flag for whether or not it supports RESIDENCY management. It should kick in if:

- The underlying allocation is not suballocated (there should be a chunk member of vkd3d_device_allocation).
- The resource or heap is DEVICE_LOCAL.
- The EXT is supported.

This can be assigned at resource and heap creation.

Reply:
Addressed this centrally at 9f15cb280f4e86f45fe4aa338774d57b9a0d3529 - please let me know if this wasn't what you had in mind.  Thanks.